ISPN-16279 Non blocking methods for Query API

wburns opened the pull request:
* Add reactive list method

https://issues.redhat.com/browse/ISPN-16279

wburns commented on client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/embedded/RemoteQueryAccessTest.java:
```diff
@@ -73,6 +77,65 @@ public void remoteQueries() {
+            .blockingGet();
+      if (maxReturn == 1) {
+         assertThat(list).extracting("name")
+               .containsExactly("The Secret of Monkey Island");
```
This is in the opposite order than expected in the test... it returns `Monkey Island 2: LeChuck's Revenge` first instead. Is this because the order is not guaranteed by the query? Maybe I should just change the assertion. @fax4ever please confirm.

fax4ever replied:
Exactly the order is not guaranteed if you don't put an `order by` clause to the Iclke text.

wburns replied:
K, just seemed odd to me that the test previously was asserting an exact ordering, but then when I limit the returned entry count it didn't have the same. I will change the test though to assume it always returns the second one though when a limit is applied.

pruivo commented on commons/all/src/main/java/org/infinispan/commons/api/query/Query.java:
```diff
@@ -23,6 +27,22 @@ public interface Query<T> extends Iterable<T> {
+    * @return a Publisher that when subscribed will query
+    */
+   @Experimental
+   default Publisher<T> listAsync(int batchSize) {
```
nitpicking: I think the method should be named `publish()` or something similar. It doesn't return a list :)

wburns replied:
Sure, sounds good 👍

fax4ever reviewed (changes requested):
Hi @wburns. Thanks for the contribution.
It seems that we have some failures on the CI.

I don’t know if we can introduce those changes more gradually.
If we could limit the scope of those changes only to the remote client code, this will help me a lot.
Probably I’m a too cautious guy…. :P
Sorry for that

wburns commented on commons/all/src/main/java/org/infinispan/commons/api/query/Query.java:
```diff
@@ -33,6 +53,17 @@ public interface Query<T> extends Iterable<T> {
+
    /**
     * Executes a data modifying statement (typically a DELETE) that does not return results; instead it returns the
     * count of affected entries. This method cannot be used to execute a SELECT.
```
Yeah, we can add that one as well 👍

wburns commented on commons/all/src/main/java/org/infinispan/commons/api/query/impl/QuerySubscription.java:
```diff
@@ -0,0 +1,63 @@
+   private int currentOffset;
+
+   protected QuerySubscription(int batchSize, Query<E> query) {
+      super(batchSize, () -> null, query);
```
I had debated this, it honestly shouldn't matter, but I can do something.

wburns commented on commons/all/src/main/java/org/infinispan/commons/api/query/impl/QuerySubscription.java:
```diff
@@ -0,0 +1,63 @@
+   private final long initialOffset;
+   private final int maxResults;
+
+   private int currentOffset;
```
Good point, investigating

wburns commented:
> Hi @wburns. Thanks for the contribution. It seems that we have some failures on the CI.
> 
> I don’t know if we can introduce those changes more gradually. If we could limit the scope of those changes only to the remote client code, this will help me a lot. Probably I’m a too cautious guy…. :P Sorry for that

All of the CI failures are solely because of a nested exception is now wrapped is all... nothing actually is broken with the changes. I honestly think doing it slower is not a good idea as we need to get this in for the next build. The API is also marked as Experimental so I see now issue honestly.

And the code here is actually quite trivial and very compact. All it requires is for each of the various Query impls to just implement executeAsync to work transparently and that is it.

wburns commented:
And I am pretty sure fixing https://github.com/infinispan/infinispan/pull/12693/files/657f3c3166527e870e27cd905de45d3eee643642#diff-af35ab757f495b925ba8c5868a6d7bc82801bda6b525f77c22190c77da17eadc will fix the test failures.

fax4ever commented:
> > Hi @wburns. Thanks for the contribution. It seems that we have some failures on the CI.
> > I don’t know if we can introduce those changes more gradually. If we could limit the scope of those changes only to the remote client code, this will help me a lot. Probably I’m a too cautious guy…. :P Sorry for that
> 
> All of the CI failures are solely because of a nested exception is now wrapped is all... nothing actually is broken with the changes. I honestly think doing it slower is not a good idea as we need to get this in for the next build. The API is also marked as Experimental so I see now issue honestly.
> 
> And the code here is actually quite trivial and very compact. All it requires is for each of the various Query impls to just implement executeAsync to work transparently and that is it.

Hi @wburns, I'll be happy to have a look after the CI is fixed. Thank you. Sorry if I'm too scary about changes on queries :P

wburns commented:
Addressed comments and confirmed tests pass that were failing before.

pruivo reviewed (approved):
It looks good to me.

fax4ever commented:
> Addressed comments and confirmed tests pass that were failing before.

You're right. Running the test locally, all seems to working. I don't know if we can have some impact on performances...

wburns commented:
> > Addressed comments and confirmed tests pass that were failing before.
> 
> You're right. Running the test locally, all seems to working. I don't know if we can have some impact on performances...

This PR changes nothing with the underlying impl, this is API only. This basically does the pagination for the user, so they don't have to do multiple queries and changing the offset themselves. You could try comparing the perf between user pagination and this if you wanted though, but this is also non blocking, so it isn't apples to apples either.

fax4ever commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/query/RemoteQuery.java:
```diff
@@ -53,13 +53,14 @@ public void resetQuery() {
    }
 
    @Override
+   @SuppressWarnings("checkstyle:ForbiddenMethod")
```
Do we still need this annotation?

wburns replied:
Nope!

fax4ever reviewed (approved):
Just a very marginal question. For the rest +1 to merge it

fax4ever commented:
My guess is that `
org.infinispan.lock.ConsistentReliabilitySplitBrainTest.ThreadLeakChecker` is unrelated. Merged, thanks @wburns.
